File: OC/GEPhotoPreviewControllerDemo/GEPhotoPreviewControllerDemo/ViewController.m
//
//  ViewController.m
//  GEPhotoPreviewControllerDemo
//
//  Created by hjp on 17/4/18.
//  Copyright © 2017年 eric. All rights reserved.
//

#import "ViewController.h"
#import "GEPhotoPreviewController.h"
#import "UIScreen+GEAdd.h"

@interface ViewController ()<GEPhotoPreviewControllerDelegate,GEPhotoPreviewControllerAnimationDataSource>

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    
   
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    GEPreviewInnerObject *photoObject = [[GEPreviewInnerObject alloc] init];
    
    photoObject.praceholder = @"yx_mm_gpres";
    photoObject.thumbnailUrl = @"https://avatars1.githubusercontent.com/u/10254305?v=3&u=373d7bf218f00f8a3d55b4cec1c2fda40babadbd&s=400";
    photoObject.orignalUrl = @"https://avatars1.githubusercontent.com/u/10254305?v=3&u=373d7bf218f00f8a3d55b4cec1c2fda40babadbd&s=400";
    
    photoObject.photoType = GEPreviewInnerObjectTypeNet;
    
    
    GEPreviewInnerObject *photoObject1 = [[GEPreviewInnerObject alloc] init];
    
    photoObject1.praceholder = @"yx_mm_gpres";
    photoObject1.thumbnailUrl = @"https://avatars1.githubusercontent.com/u/10254305?v=3&u=373d7bf218f00f8a3d55b4cec1c2fda40babadbd&s=400";
    photoObject1.orignalUrl = @"https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1492576419788&di=95572b657eadc0346e5935d85f34cd48&imgtype=0&src=http%3A%2F%2Fs15.sinaimg.cn%2Fmw690%2F003vNYwKgy6EdhlrawKae%26690";
    
    photoObject1.photoType = GEPreviewInnerObjectTypeNet;
    
    GEPreviewInnerObject *photoObject2 = [[GEPreviewInnerObject alloc] init];

    photoObject2.praceholder = @"yx_mm_gpres";
    photoObject2.thumbnailUrl = @"https://avatars1.githubusercontent.com/u/10254305?v=3&u=373d7bf218f00f8a3d55b4cec1c2fda40babadbd&s=400";
    photoObject2.orignalUrl = @"https://imgsa.baidu.com/baike/c0%3Dbaike180%2C5%2C5%2C180%2C60/sign=ca5abb5b7bf0f736ccf344536b3cd87c/29381f30e924b899c83ff41c6d061d950a7bf697.jpg";
    
    photoObject2.photoType = GEPreviewInnerObjectTypeNet;
    
    
    GEPhotoPreviewController *photoPreview = [GEPhotoPreviewController previewByPhotoObjects:@[photoObject,photoObject1,photoObject2] selectIndex:1 ];
    
    photoPreview.delegate = self;
    photoPreview.dataSource = self;
    
    [self.navigationController presentViewController:photoPreview animated:YES completion:nil];
}

#pragma mark - GEPhotoPreviewControllerAnimationDataSource
- (CGRect)startRectForPhotoPreviewAnimation:(NSIndexPath *)animateIndexPath
{
    CGRect bounds = [UIScreen bounds];
    return CGRectMake(bounds.size.width * 0.4, bounds.size.height * 0.4, 50, 50);
}
- (void)photoPreview:(GEPhotoPreviewController *)photoBrowser tapImageView:(GEPreviewInnerObject *)photoObject item:(NSInteger)item
{
//    [self dismissViewControllerAnimated:YES completion:nil];

}
- (void)photoPreview:(GEPhotoPreviewController *)photoBrowser longPressImageViewObject:(GEPreviewInnerObject *)photoObject image:(UIImage *)image item:(NSInteger)item
{
    NSLog(@"长按");
}


@end

File: Swift/Swift3基础/6-Closure使用/3-Closure的循环引用/6-Closure/ViewController.swift
//
//  ViewController.swift
//  6-Closure
//
//  Created by 黄继平 on 2016/11/17.
//  Copyright © 2016年 PJStudio. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    
    /** 网络工具*/
    private var httpTools : HTTPTools?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        httpTools = HTTPTools()

    }

    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        

        // 逃逸闭包会对其中的对象做强引用,这也是为什么闭包中我们使用当前控制的属性时,需要强制写self,而其他方法中我们可以省略self直接调用属性
        
        // 内存分析
        /*
                    self
                    /   ^
                   /     \
                  v       \
            httpTools      closure
         
         上图可见self强引用者httpTools(默认是强引用),而closure强引用者self,但是,并不构成"死循环",所以不存在循环引用问题;
         如果我们在httpTools中强引用一下closure,那么就构成死循环了!!!
         */
        
        
        //        httpTools?.loadData { (jsonData) in
        //            
        //            print("得到数据:\(jsonData)")
        //            self.view.backgroundColor = UIColor.red
        //            
        //        }
        
        /* 当httpTools引用这closure时,就构成死循环了,也就是所谓循环引用
         
                    self
                   /   ^
                  /     \
                 v       \
         httpTools  --> closure
         
         */
        
        /*
            为了避免循环引用,我们有几个思路:
            1.弱化self对httpTools的引用;
            2.弱化httpTools对closure的引用
            3.弱化closure对self的引用
            
            具体弱化哪个环节,需要根据具体情境!!💕
         */
        
        // 此处,由于self属于shortLived环节,所以,我们选择弱化closure对self的引用!
        /*
         方案1:
         weakSolution1()
        
         方案2:
         weakSolution2()

         方案3:
         weakSolution3()

         */
        
    }
    
   
    private func weakSolution1(){
    
        weak var weakSelf = self
        
        httpTools?.loadData({ (jsonData) in
            
            print("得到返回数据:\(String(describing: jsonData)),弱引用方案1")
            
            weakSelf?.view.backgroundColor = UIColor.green
        })
        
        /*
         请求数据,当前线程:<NSThread: 0x610000260e40>{number = 3, name = (null)}
         得到返回数据:Optional("图片一张"),弱引用方案1
         */
    }
    
    private func weakSolution2()->(){
        httpTools?.loadData{ [weak self] (jsonData) in
            
            print("得到返回数据:\(String(describing: jsonData)),弱引用方案2")
            
            // 注意,此处self已经是可选的了!
            self?.view.backgroundColor = UIColor.red
        }
        /*
         请求数据,当前线程:<NSThread: 0x600000075b80>{number = 3, name = (null)}
         得到返回数据:Optional("图片一张"),弱引用方案2
         
         */

    }
    
    private func weakSolution3()->(){
    
        httpTools?.loadData{ [unowned self] (jsonData) in
            
            print("得到返回数据:\(String(describing: jsonData)),弱引用方案3")
            
            // 注意,此处self是不可选的,此处unowned类似于OC中的__unsafe__unretain修饰符
            // 表示,closure不会对self强引用(引用计数未加1️⃣),但是,当self deinit时,closure使用self指针访问的内存不会自动转移到nil(0x0),而是继续访问被释放的内存,不安全!⚠️
            self.view.backgroundColor = UIColor.orange
        }
        /*
         请求数据,当前线程:<NSThread: 0x60000026a240>{number = 3, name = (null)}
         得到返回数据:Optional("图片一张"),弱引用方案3
         */

    }
}

